packages/appium/lib/appium.js

```diff
@@ -241,6 +252,330 @@ class AppiumDriver extends DriverCore {
+      // the client
+      proxyClient.on('message', async (/** @type {Buffer} */ data) => {
+        const logData = _.truncate(data.toString('utf8'), {length: MAX_LOG_BODY_LENGTH});
+        this.log.debug(
+          `<-- BIDI Received data from proxied bidi socket, sending to client. Data: ${logData}`,
+        );
+        await send(data);
+      });
+
+      // If the upstream socket server closes the connection, should close the connection to the
+      // client as well
+      proxyClient.on('close', (code, reason) => {
+        this.log.debug(
+          `Upstream bidi socket closed connection (code ${code}, reason: '${reason}'). ` +
+            `Closing proxy connection to client`,
+        );
+        ws.close(code, reason);
+      });
+
+      proxyClient.on('error', (err) => {
+        this.log.error(`Got error on upstream bidi socket connection: ${err}`);
+      });
+    }
+  }
+
+  /**
+   * Set up handlers on the bidi socket connection to the client
+   *
+   * @param {import('ws').WebSocket} ws - the websocket connection to/from the client
+   * @param {import('ws').WebSocket | null} proxyClient - the websocket connection to/from the
+   * upstream socket (the one we're proxying to/from, if we're proxying)
+   * @param {(data: string | Buffer) => Promise<void>} send - a method used to send data to the
+   * client
+   * @param {((data: string | Buffer) => Promise<void>) | null} sendToProxy - a method used to send data to the
+   * upstream socket
+   * @param {import('@appium/types').ExternalDriver | AppiumDriver} bidiHandlerDriver - the driver
+   * handling the bidi commands
+   * @param {(err: Error) => void} logSocketErr - a special prefixed logger
+   */
+  initBidiSocketHandlers(ws, proxyClient, send, sendToProxy, bidiHandlerDriver, logSocketErr) {
+    ws.on('error', (err) => {
+      // Can't do much with random errors on the connection other than log them
+      logSocketErr(err);
+    });
+
+    ws.on('open', () => {
+      this.log.info('Bidi websocket connection is now open');
+    });
+
+    // Now set up handlers for the various events that might happen on the websocket connection
+    // coming from the client
+    // First is incoming messages from the client
+    ws.on('message', async (/** @type {Buffer} */ data) => {
+      if (proxyClient) {
+        const logData = _.truncate(data.toString('utf8'), {length: MAX_LOG_BODY_LENGTH});
+        this.log.debug(
+          `--> BIDI Received data from client, sending to upstream bidi socket. Data: ${logData}`,
+        );
+        // if we're meant to proxy to an upstream bidi socket, just do that
+        // @ts-ignore sendToProxy is never null if proxyClient is truthy, but ts doesn't know
+        // that
+        await sendToProxy(data.toString('utf8'));
+      } else {
+        const res = await this.onBidiMessage(data, bidiHandlerDriver);
+        await send(JSON.stringify(res));
```

Comment:
I assume `res` might be pretty large. Do we really want to load it all into RAM? Can it be streamed instead?

Reply:
`res` is whatever the driver has decided to return to the client. it will therefore already be in memory. i don't think this websocket protocol supports streaming.

Reply:
it supports streaming for sure via pipes, for example https://stackoverflow.com/questions/22723671/streaming-data-in-node-js-with-ws-and-websocket-stream

Currently one of major limitations of the HTTP API is that we cannot stream big files, e.g. video recordings to the client, because these must be first fully stored in the memory